File: order-edit-wrong-balance-direction/node/flag-wrong-balance-direction.js
/**
 * Find Medusa v2 orders where a captured payment plus a pending or
 * confirmed order edit reports the balance direction backwards, refund
 * owed reported as collect, or the reverse. This is not auto-fixable:
 * it is a computed field bug in Medusa core (GitHub issues #13068,
 * #13067), not corrupted data. DRY_RUN=true (default) only reports the
 * affected orders. Only when DRY_RUN=false and a human has reviewed the
 * direction does the script add an internal_note to the order edit,
 * never a capture or refund call.
 */
import { pathToFileURL } from "node:url";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const ORDER_FIELDS = "id,display_id,status,*summary,*items,*order_change";
const EDIT_CHANGE_STATUSES = new Set(["requested", "confirmed"]);

const NOTE_TEMPLATE = (direction, pendingDifference) =>
  `Suspected reversed refund-direction bug (medusajs/medusa#13068) -- ` +
  `verify manually before force-confirming or capturing/refunding payment. ` +
  `recomputed_direction=${direction} pending_difference=${pendingDifference}`;

export function hasRelevantEdit(order) {
  const change = order.order_change || {};
  return change.change_type === "edit" && EDIT_CHANGE_STATUSES.has(change.status);
}

export function decideBalanceAction(currentOrderTotal, paidTotal) {
  // Pure: no I/O. pending_difference semantics per Medusa OrderSummary:
  // current_order_total - paid_total. Negative means refund owed to the
  // customer, positive means more is owed by the customer.
  const diff = Number(currentOrderTotal) - Number(paidTotal);
  if (diff === 0) return { pendingDifference: 0, direction: "none" };
  return diff < 0
    ? { pendingDifference: diff, direction: "refund" }
    : { pendingDifference: diff, direction: "collect" };
}

export function reportedDirection(order) {
  // Read the direction the app/UI is currently using, from the order's
  // own summary.pending_difference, the exact field the bug can flip.
  const summary = order.summary || {};
  const reported = summary.pending_difference;
  if (reported == null) return null;
  if (reported === 0) return "none";
  return reported < 0 ? "refund" : "collect";
}

async function login() {
  const { default: Medusa } = await import("@medusajs/js-sdk");
  const sdk = new Medusa({ baseUrl: BASE_URL, auth: { type: "jwt" } });
  await sdk.auth.login("user", "emailpass", { email: EMAIL, password: PASSWORD });
  return sdk;
}

async function listPaidEditedOrders(sdk) {
  const out = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const body = await sdk.admin.order.list({ fields: ORDER_FIELDS, limit, offset });
    for (const o of body.orders) {
      const paidTotal = o.summary?.paid_total || 0;
      if (paidTotal > 0 && hasRelevantEdit(o)) out.push(o);
    }
    offset += limit;
    if (offset >= body.count) return out;
  }
}

async function addInternalNote(sdk, orderEditId, note) {
  return sdk.client.fetch(`/admin/order-edits/${orderEditId}`, {
    method: "POST",
    body: { internal_note: note },
  });
}

export async function run() {
  const sdk = await login();
  const orders = await listPaidEditedOrders(sdk);

  const flagged = [];
  for (const order of orders) {
    const summary = order.summary || {};
    const expected = decideBalanceAction(summary.current_order_total || 0, summary.paid_total || 0);
    const reported = reportedDirection(order);
    if (reported !== null && reported !== expected.direction) {
      flagged.push([order, expected, reported]);
    }
  }

  if (flagged.length === 0) {
    console.log(`No direction mismatches found across ${orders.length} paid, edited order(s).`);
    return;
  }

  for (const [order, expected, reported] of flagged) {
    const change = order.order_change || {};
    const orderEditId = change.id;
    const summary = order.summary || {};
    console.warn(
      `Order ${order.id} (display #${order.display_id}): reported=${reported} expected=${expected.direction} ` +
      `paid_total=${summary.paid_total} current_order_total=${summary.current_order_total} ` +
      `pending_difference=${expected.pendingDifference}. ${DRY_RUN ? "Would report only" : "Reported, adding internal note"}`
    );
    if (!DRY_RUN && orderEditId) {
      await addInternalNote(sdk, orderEditId, NOTE_TEMPLATE(expected.direction, expected.pendingDifference));
    }
  }

  console.log(`Done. ${flagged.length} order(s) flagged with a mismatched balance direction.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: order-edit-wrong-balance-direction/python/flag_wrong_balance_direction.py
"""Find Medusa v2 orders where a captured payment plus a pending or
confirmed order edit reports the balance direction backwards, refund
owed reported as collect, or the reverse. This is not auto-fixable: it
is a computed field bug in Medusa core (GitHub issues #13068, #13067),
not corrupted data. DRY_RUN=true (default) only reports the affected
orders. Only when DRY_RUN=false and a human has reviewed the direction
does the script add an internal_note to the order edit, never a
capture or refund call.
"""
import os
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_wrong_balance_direction")

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDER_FIELDS = "id,display_id,status,*summary,*items,*order_change"
EDIT_CHANGE_STATUSES = {"requested", "confirmed"}

NOTE_TEMPLATE = (
    "Suspected reversed refund-direction bug (medusajs/medusa#13068) "
    "-- verify manually before force-confirming or capturing/refunding payment. "
    "recomputed_direction={direction} pending_difference={pending_difference}"
)


def get_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def has_relevant_edit(order):
    change = order.get("order_change") or {}
    return (
        change.get("change_type") == "edit"
        and change.get("status") in EDIT_CHANGE_STATUSES
    )


def list_paid_edited_orders(token):
    headers = {"Authorization": f"Bearer {token}"}
    out, offset, limit = [], 0, 100
    while True:
        r = requests.get(
            f"{BASE_URL}/admin/orders",
            params={"fields": ORDER_FIELDS, "limit": limit, "offset": offset},
            headers=headers,
            timeout=30,
        )
        r.raise_for_status()
        body = r.json()
        for o in body["orders"]:
            summary = o.get("summary") or {}
            if summary.get("paid_total", 0) > 0 and has_relevant_edit(o):
                out.append(o)
        offset += limit
        if offset >= body["count"]:
            return out


def decide_balance_action(current_order_total, paid_total):
    """Pure: no I/O. pending_difference semantics per Medusa OrderSummary:
    current_order_total - paid_total. Negative means refund owed to the
    customer, positive means more is owed by the customer."""
    diff = float(current_order_total) - float(paid_total)
    if diff == 0:
        return {"pendingDifference": 0.0, "direction": "none"}
    if diff < 0:
        return {"pendingDifference": diff, "direction": "refund"}
    return {"pendingDifference": diff, "direction": "collect"}


def reported_direction(order):
    """Read the direction the app/UI is currently using, from the order's
    own summary.pending_difference, the exact field the bug can flip."""
    summary = order.get("summary") or {}
    reported = summary.get("pending_difference")
    if reported is None:
        return None
    if reported == 0:
        return "none"
    return "refund" if reported < 0 else "collect"


def add_internal_note(token, order_edit_id, note):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.post(
        f"{BASE_URL}/admin/order-edits/{order_edit_id}",
        json={"internal_note": note},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    token = get_token()
    orders = list_paid_edited_orders(token)

    flagged = []
    for order in orders:
        summary = order.get("summary") or {}
        expected = decide_balance_action(
            summary.get("current_order_total", 0), summary.get("paid_total", 0)
        )
        reported = reported_direction(order)
        if reported is not None and reported != expected["direction"]:
            flagged.append((order, expected, reported))

    if not flagged:
        log.info("No direction mismatches found across %d paid, edited order(s).", len(orders))
        return

    for order, expected, reported in flagged:
        change = order.get("order_change") or {}
        order_edit_id = change.get("id")
        summary = order.get("summary") or {}
        log.warning(
            "Order %s (display #%s): reported=%s expected=%s paid_total=%s "
            "current_order_total=%s pending_difference=%s. %s",
            order["id"], order.get("display_id"), reported, expected["direction"],
            summary.get("paid_total"), summary.get("current_order_total"),
            expected["pendingDifference"],
            "Would report only" if DRY_RUN else "Reported, adding internal note",
        )
        if not DRY_RUN and order_edit_id:
            note = NOTE_TEMPLATE.format(
                direction=expected["direction"],
                pending_difference=expected["pendingDifference"],
            )
            add_internal_note(token, order_edit_id, note)

    log.info("Done. %d order(s) flagged with a mismatched balance direction.", len(flagged))


if __name__ == "__main__":
    run()
